pkg/provisioner/fss/fss.go

```diff
@@ -28,248 +28,388 @@ import (
-	{
-		ctx, cancel := context.WithTimeout(ctx, fsp.client.Timeout())
-		defer cancel()
-		resp, err := fsp.client.FSS().CreateExport(ctx, filestorage.CreateExportRequest{
-			CreateExportDetails: filestorage.CreateExportDetails{
-				ExportSetId:  target.ExportSetId,
-				FileSystemId: &fsID,
-				Path:         common.String("/" + fsID),
-			},
-		})
-		if err != nil {
-			logger.With(zap.Error(err)).Error("Failed to create export.")
-			return nil, err
-		}
-		exportSetID = *resp.Export.Id
-		logger = logger.With("exportSetID", exportSetID)
-	}
-
+	// Ensure MountTarget required fields are set.
 	if len(target.PrivateIpIds) == 0 {
-		logger.With("targetID", *target.Id).Error("Could not find any Private IPs for Mount Target.")
-		return nil, errors.Errorf("failed to find server IDs associated with the Mount Target with OCID %q", *target.Id)
+		logger.Error("Failed to find private IPs associated with the Mount Target")
+		return nil, errors.Errorf("mount target has no associated private IPs")
+	}
+	if target.ExportSetId == nil {
+		logger.Error("Mount target has no export set associated with it")
+		return nil, errors.Errorf("mount target has no export set associated with it")
 	}
 
-	// Get PrivateIP.
-	var serverIP string
+	// Randomly select a MountTarget IP address to attach to.
+	var ip string
 	{
-		ctx, cancel := context.WithTimeout(ctx, fsp.client.Timeout())
-		defer cancel()
+
 		id := target.PrivateIpIds[rand.Int()%len(target.PrivateIpIds)]
-		getPrivateIPResponse, err := fsp.client.VCN().GetPrivateIp(ctx, core.GetPrivateIpRequest{
-			PrivateIpId: &id,
-		})
+		logger = logger.With("privateIPID", id)
+		resp, err := fsp.client.VCN().GetPrivateIp(ctx, core.GetPrivateIpRequest{PrivateIpId: &id})
 		if err != nil {
-			logger.With(zap.Error(err), "privateIPID", id).Errorf("Failed to retrieve IP address for mount target privateIpID=%q: %v", id, err)
+			logger.With(zap.Error(err)).Error("Failed to retrieve IP address for mount target")
 			return nil, err
 		}
-		serverIP = *getPrivateIPResponse.PrivateIp.IpAddress
+		if resp.PrivateIp.IpAddress == nil {
+			logger.Error("PrivateIp has no IpAddress")
+			return nil, errors.Errorf("PrivateIp %q associated with MountTarget %q has no IpAddress", id, mtID)
+		}
+		ip = *resp.PrivateIp.IpAddress
 	}
+	logger = logger.With("privateIP", ip)
 
-	region, ok := os.LookupEnv("OCI_SHORT_REGION")
-	if !ok {
-		metadata, err := fsp.metadata.Get()
-		if err != nil {
-			return nil, err
-		}
-		region = metadata.Region
+	logger.Info("Creating FileSystem")
+	fs, err := fsp.getOrCreateFileSystem(ctx, logger, *ad.Name, fsDisplayName)
+	if err != nil {
+		return nil, err
 	}
+	logger = logger.With("fileSystemID", *fs.Id)
 
-	logger.With("privateIP", serverIP).Info("Selected Mount Target Private IP.")
+	logger.Info("Creating Export")
+	export, err := fsp.getOrCreateExport(ctx, logger, *fs.Id, *target.ExportSetId)
+	if err != nil {
+		logger.With(zap.Error(err)).Error("Failed to create export.")
```

Comment:
Nitpick: Inconsistent punctation in error messages, end with full stop or no?

Reply:
Good catch - my personal preference is to only use full stops when there's multiple sentances in the message.